### ISSUE TITLE
Warn on Invalid Exit Codes

### ISSUE THREAD

**Issue by eatnumber1** [koalaman/shellcheck#1228]
#### For new checks and feature suggestions
- [Y] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [Y] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

It's quite poorly documented, but any exit codes outside the range of 0 - 125 are unsafe / incorrect to use. See [this stackoverflow post](https://stackoverflow.com/questions/23262887/waitpid-wifexited-returning-0-although-child-exited-normally/23272329#23272329) for info about the way that the return value of exit is actually used, wherein you can infer the valid exit codes. Basically, if you use anything above 125, you can step on either the coredump bit or the bits for the signal number / flags. Also, the exit code is treated as an unsigned integer, so -1 is invalid.

ShellCheck should warn if it can deduce that you're exiting with an invalid exit code, e.g. `exit 255`.

**Comment by koalaman** [comment]
From the line `8-15   low 8 bits of value passed to _exit/exit or returned by main,` I would infer that the valid exit codes are 0-255. Why would it step on the core dump bit, especially when that bit is lower than the exit code?

**Comment by jleffler** [comment]
You may have misunderstood the cross-referenced answer in SO — or are using it in ways that are not intended by the answer.

Exit statuses outside the range 0..255 are dubious, though `exit -1` normally equates to `exit 255` since only the 8 lower-order bits are used when the status passed to the `exit()` system call are relayed to the waiting parent program; other negative numbers are also translated by taking the least significant 8 bits of the value and treating it (for most purposes) as an unsigned 8-bit number.

POSIX does define some exit statuses for failures to execute programs in various ways (126, 127), but those statuses were available for any use long before POSIX got anywhere near the scene and remain valid.  There is a small chance that such an exit status will cause confusion, but it is not illegal for any program to exit with status 126 or 127.  (Note that an exit status 1..125 should be used to report problems during word expansion or I/O redirection.)

Bash does encode the exit status of a process that died from a signal as '128 + signal-number'.  For example, if a Bash script ends with a program that dies from signal 6 (e.g. `suicide -s 6` — where `suicide` is a program that kills itself with the specified signal), Bash will relay an exit status of 134 (128+6) to the program that invoked the script. But that simply reuses some exit statuses which have decades of perfectly respectable use for other purposes, albeit a use that was seldom exercised.  [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_08_02) only demands that the reported exit status of a command that dies from a signal shall be greater than 128 — Bash has chosen a specific (and reasonable) interpretation of that requirement.

There are programs that report multiple statuses as parts of a bit mask via the exit status — not necessarily reporting errors, just what it found while working (`sunifdef` is one such program).

At the level of the `wait`-family of system calls, the bits used to relay signal and core dump information are wholly separate from the bits used to relay exit status.  Bash has chosen to try mapping the two sets of 8 bits into one 8-bit quantity; it isn't an entirely happy coexistence.

See also:

* [`system()`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/system.html)
* [`_exit()`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/_exit.html)
* [`exit()`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/exit.html)
* [`pclose()`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/pclose.html)
* [`waitpid()`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/waitpid.html)

**Comment by vapier** [comment]
i agree that shellcheck should issue a warning whenever a script tries to use 126 or 127 (relevant [POSIX documentation](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_08_02)).

for all dialects, it should warn when using values below 0 or greater than 255 since only the low 8 bits are available via most interfaces, and POSIX only talks about the low 8 bits for shell code.  relevant docs:
* [exit()](http://pubs.opengroup.org/onlinepubs/9699919799/functions/exit.html)
* [Exit Status for Commands](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_08_02)
* [exit command](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#exit)
* [return command](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#return)

for the bash dialect, it should warn when using values between [128,255] since bash will use them to indicate signal exits.  i think it'd be OK to warn for all dialects here since other shells seem to do the same thing (at least dash does).

shellcheck already warns about `return` values being out of range of [0,255] via SC2152.

no one is saying that a script using any exit status is "illegal".  however, the point of shellcheck isn't to catch only "illegal" code and ignore the rest ... it's meant to be a comprehensive linter providing warnings & suggestions about many practices that people don't realize are (arguably) bad.  so in that regard, shellcheck warning about this is perfectly appropriate imo.  plus, if you don't like it, like every other lint category out there, you can simply disable it in your script via a shellcheck directive or via the command line `-e` option.

**Comment by koalaman** [comment]
@vapier I still don't follow. That link says "Applications that invoke utilities without using the shell should use these exit status values to report similar errors."

While there's still a shell involved, doesn't that mean that e.g. `if ! type ffmpeg; then exit 127; fi` would be the correct thing to do?

**Comment by vapier** [comment]
i think it depends exactly what the script is accomplishing.  if it's supposed to be a transparent wrapper, you could probably argue that 127 is OK.  but generally exiting 126/127 is incorrect, and i think the scenario you highlight is uncommon enough to make it into a warning/suggestion that the user would, in a valid but still extraordinary circumstance, disable in the single usage.

**Comment by Gandalf-** [comment]
This was fixed by https://github.com/koalaman/shellcheck/commit/cb76951ad23616b971493674eea0170864bddbc4 

The commit checks for negative values and values greater than 255

This issue can be closed

**Comment by vapier** [comment]
that is a good start, but as documented above, there's more to it.  we should warn on [126,255] exits due using a distinct category.

**Comment by wileyhy** [comment]
I for one find `exit $LINENO` to be particularly useful during debugging.
